### PR TITLE
fix(controller): add handling for invalid cursor edge case

### PR DESCRIPTION
**Pull request by salmoneatenbybear**
### Fixes the error raised by sentry.io bot about the invalid cursor in #5199 ( the main pr ) for the fix of #5195 

- Added validation for `params[:after]` and `params[:before]` to ensure cursors are in the correct format.
- Handled `InvalidCursorError` with a fallback to the first page and logging for debugging.
- This fixes potential issues with malformed or invalid cursor inputs and improves robustness of the recent circuits view.

## Summary by CodeRabbit

- **Bug Fixes**
	- Improved error handling for pagination by introducing a fallback mechanism for invalid cursor errors.
  
- **Refactor**
	- Updated the logic for constructing project queries to improve maintainability.

- **Tests**
	- Added test coverage for handling invalid cursor parameters in the controller.

### CODE REVIEW

**Comment by salmoneatenbybear** [comment]
@tachyons please review if this solved the issue raised by sentry on the cursor based pagination pr. 

would be happy to make changes.

**Comment by tachyons** [comment]
Could you add spec for this change ?

**Comment by salmoneatenbybear** [comment]
> Could you add spec for this change ?

sure absolutely.

File: app/controllers/circuitverse_controller.rb [hunk 1]
```diff
@@ -4,18 +4,33 @@ class CircuitverseController < ApplicationController
...
+      @projects_paginator = Project.select(:id, :author_id, :image_preview, :name, :slug)
+                                   .public_and_not_forked
+                                   .where.not(image_preview: "default.png")
+                                   .includes(:author)
+                                   .cursor_paginate(
+                                     order: { id: :desc },
+                                     limit: Project.per_page
+                                   )
+      @projects_page = @projects_paginator.fetch
+    end
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply app/controllers/circuitverse_controller.rb:32]: should i do this @tachyons or is this not needed ?

**Comment by salmoneatenbybear** [comment]
@tachyons added spec for the invalid cursor handling

**Comment by salmoneatenbybear** [comment]
is test for the before parameter needed ?

File: app/controllers/circuitverse_controller.rb [hunk 2]
```diff
@@ -32,18 +47,14 @@ def index
...
-                   img: "examples/FlipFlop_n.jpg" },
-                 { name: "ALU 74LS181 by Ananth Shreekumar", id: "users/126/projects/252",
-                   img: "examples/ALU_n.jpg" }]
+    @examples = [
+      { name: "Full Adder from 2-Half Adders", id: "users/3/projects/247", img: "examples/fullAdder_n.jpg" },
+      { name: "16 Bit ripple carry adder", id: "users/3/projects/248", img: "examples/RippleCarry_n.jpg" },
+      { name: "Asynchronous Counter", id: "users/3/projects/249", img: "examples/AsyncCounter_n.jpg" },
+      { name: "Keyboard", id: "users/3/projects/250", img: "examples/Keyboard_n.jpg" },
+      { name: "FlipFlop", id: "users/3/projects/251", img: "examples/FlipFlop_n.jpg" },
+      { name: "ALU 74LS181 by Ananth Shreekumar", id: "users/126/projects/252", img: "examples/ALU_n.jpg" }
```
Comment (tachyons) [inline app/controllers/circuitverse_controller.rb:56]: Unrelated

Comment (salmoneatenbybear) [reply app/controllers/circuitverse_controller.rb:56]: let me revert this change

File: app/controllers/circuitverse_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (tachyons) [reply app/controllers/circuitverse_controller.rb:32]: Yes

[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply app/controllers/circuitverse_controller.rb:32]: noted

[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply app/controllers/circuitverse_controller.rb]: done

**Comment by salmoneatenbybear** [comment]
@tachyons added the spec for handling after aswell as before for the invalid cursor, refactored the controller to make the code cleaner as suggested by @coderrabbitai, removed unrelated changes to the array formatting, ran the test for the file and esnured its working, also ensured the recent circuits are working as intended, below are to verify the same. 

image



https://github.com/user-attachments/assets/c14dc006-32fc-4548-b328-1b79f6dda436

**Comment by salmoneatenbybear** [comment]
Thankyou @tachyons let me know if any updates are needed for this

**Pull request merged**